### ISSUE TITLE
Workers apparently not achieving the expected concurrency

### ISSUE THREAD

**Issue by khurtado** [matz-e/lobster#179]
This is just to keep track of @klannon issue presented at  the Aug 27 Thursday meeting until we completely we understand the issue.

The following lobster task:
http://www.crc.nd.edu/~klannon/ttH_13TeV_debug2/

shows plenty of workers on idle even though there is still work to do.
http://www.crc.nd.edu/~klannon/ttH_13TeV_debug2/workers-plot.png

The question is whether this has something to do with either work_queue or lobster.

This is what has been looked into so far:

In [1], I think this is a lobster thing. In short, no more jobs are created because all jobits are distributed within the jobs that are already running, so nothing additional for work_queue to run.

In [2], the database in [0] shows that jobs get to process 256 jobits at most (without considering merge jobs), even though initial number of jobits is up to 814. One of the reasons can be related to issue #178, where we are apparently only appending the first input file detected from parameters.job.

********************************

[0]
`````
sqlite> select id, jobits, jobits_processed, events_read from jobs;
1|814|251|49574
2|814|252|49484
3|814|100|19579
4|814|152|30169
5|814|252|49681
6|814|256|49842
7|814|250|49913
8|814|14|2701
9|814|241|47478
10|814|254|49845
11|814|256|49887
12|814|232|45867
13|814|86|17033
14|814|251|49628
15|814|250|49442
16|814|183|36077
17|69|69|13621
18|800|12|2353
19|788|230|45439
...
78|23|23|4541
79|188|142|28156
80|406|154|30364
81|252|252|49827
82|46|26|5084
83|20|20|3967
84|4347|4347|0
````

[1] Here is my first reply to it via email:
--------
I think this is not a work_queue issue but a lobster thing...

I looked for the time when Kevin increased the number of workers in the lobster logfile (12:52 in this plot http://www.crc.nd.edu/~klannon/ttH_13TeV_debug2/workers-plot.png) and saw the following:

````
2015-08-22 12:47:51,751 [ INFO] - ...Lobster-1.2-py2.7.egg/lobster/core.py  221: 1 out of 1 workers busy; 2 jobs running, 0 waiting; 850 jobits left
2015-08-22 12:52:51,194 [ INFO] - ...Lobster-1.2-py2.7.egg/lobster/core.py  221: 1 out of 11 workers busy; 2 jobs running, 0 waiting; 850 jobits left
````
I think the fact there are 0 jobs waiting means Lobster hasn't created any more jobs for work_queue to run. A possible reason would be that the 850 jobits left are already distributed in the 2 jobs running.
When I looked into the database, I confirmed the 2 jobs running where number 25 and 79, which had 662 and 188 jobits assigned respectively, so that's the reason no more jobs can be immediately created and sent to work_queue.

The reason more non-merge jobs are created after a while, is because even though jobs 25 and 79 eventually finish, they missed jobits, so new jobs are created from those missing jobits.That means at that point, you need a job to finish before creating and running another from the possible remaining jobits.
For example, Job 25 only managed to process 256 out of 662 jobits, hence Job 80 was created with the remaining 406 jobits after a while. A similar thing happened with job 79.

As for why a successful job would miss jobits, I'm not sure how the CMS framework works in this case, but I suppose one possibility is that if you have a  job reading e.g 5 files via XRootD and you have connection/handshake/etc problems after a while -but your app still managed to process some events and get an output from those- then e.g cmsRun still reports itself as successful, but its report will only include the lumis/events/files that it was able to process. Something similar would be cases where you copy your inputs and only some of them weere successfully copied for some reason, then the list of files to run is modified before running the app and lobster report the missing jobits associated with the files it couldn't copy at the time.

[2]Second email
--------

Yes, I looked earlier and also noticed the 0 jobs waiting, so nothing to send work_queue:

2015-08-22 09:38:48,750 [ INFO] - ...Lobster-1.2-py2.7.egg/lobster/core.py  221: 7 out of 25 workers busy; 15 jobs running, 0 waiting; 12815 jobits left
2015-08-22 10:06:51,569 [ INFO] - ...Lobster-1.2-py2.7.egg/lobster/core.py  221: 8 out of 26 workers busy; 11 jobs running, 0 waiting; 6404 jobits left
2015-08-22 10:56:53,212 [ INFO] - ...Lobster-1.2-py2.7.egg/lobster/core.py  221: 4 out of 14 workers busy; 5 jobs running, 0 waiting; 3038 jobits left

However, I found strange that your database never had more than 250 jobits finished per job (even though there are initially 814 jobits assigned for each), so I think we also have a bug in lobster.

Basically, we create a parameters.json file with a list of:

1. Input Files
2. Lumi sections

that we pass to the job.py wrapper, but even though we create the cmsRun with the full list of lumis, we only send 1 Input File. I will make a github issue for it describing in more detail.

**Comment by matz-e** [comment]
Is this still an issue with the fix in master?

**Comment by klannon** [comment]
I think this is fixed but need to try Serguei's workaround for my grid 
certificate issue in order to run the test.  Will report back soon.

**Comment by khurtado** [comment]
Copy/paste from @klannon latest email, to close this issue:

To close the loop, after hacking some things so that all pieces are accessing up-to-date grid certificates, I was able to run a task to completion using a fixed version of Lobster where Kenyi and Matthias addressed the problems we discussed this morning.  Running over the same inputs (although with a slightly different task sizing, but that's almost guaranteed because of the input file bug that was fixed), I was able to finish the work in about ~1 hour instead of ~7 hours.  Lobster was also able to use all the workers I gave it until I got too excited and gave it so many that it ran out of work.  Thanks to Ben, Kenyi, and Matthias for sorting this out!
